pyomo/gdp/plugins/transform_current_disjunctive_state.py

```diff
@@ -0,0 +1,297 @@
+                    "are selected: %s"
+                    % (disjunction.name, ', '.join([d.name for d in true_val]))
+                )
+            elif len(true_val) == 1:
+                # We can fix everything
+                self._reclassify_disjuncts(
+                    true_val,
+                    no_val.union(false_val),
+                    reverse_dict['_disjuncts'],
+                    disjunct_set,
+                    disjunct_containers,
+                )
+                # This disjunction is fully transformed
+                reverse_dict['_disjunctions'].add(disjunction)
+                disjunction.deactivate()
+            elif len(false_val) == num_disjuncts - 1:
+                # We can fix everything. Since we didn't hit the case above, we
+                # know that the non-False value is a None.
+                self._reclassify_disjuncts(
+                    no_val,
+                    false_val,
+                    reverse_dict['_disjuncts'],
+                    disjunct_set,
+                    disjunct_containers,
+                )
+                reverse_dict['_disjunctions'].add(disjunction)
+                disjunction.deactivate()
+            else:
+                can_transform = False
+        # It's only an 'at-least' not an 'exactly', so if everything has a value
+        # we can transform it
+        elif len(no_val) == 0:
+            self._reclassify_disjuncts(
+                true_val,
+                false_val,
+                reverse_dict['_disjuncts'],
+                disjunct_set,
+                disjunct_containers,
+            )
+            reverse_dict['_disjunctions'].add(disjunction)
+            disjunction.deactivate()
+        else:
+            can_transform = False
+        if not can_transform:
+            raise GDP_Error(
+                "Disjunction '%s' does not contain enough Disjuncts with "
+                "values in their indicator_vars to specify which Disjuncts "
+                "are True. Cannot fully transform model." % disjunction.name
+            )
+
+    def _update_transformed_disjuncts(self, disj, disjunct_set, disjunct_containers):
+        parent = disj.parent_component()
+        if parent.is_indexed():
+            if parent not in disjunct_containers:
+                disjunct_set.update(d for d in parent.values() if d.active)
+                disjunct_containers.add(parent)
+            disjunct_set.remove(disj)
+
+    def _reclassify_disjuncts(
+        self,
+        true_disjuncts,
+        false_disjuncts,
+        reverse_dict,
+        disjunct_set,
+        disjunct_containers,
+    ):
+        for disj in true_disjuncts:
+            reverse_dict[disj] = (disj.indicator_var.fixed, disj.indicator_var.value)
+            self._update_transformed_disjuncts(disj, disjunct_set, disjunct_containers)
+            parent_block = disj.parent_block()
+            parent_block.reclassify_component_type(disj, Block)
```

Comment:
I think this is actually okay--we're comparing the GDP hierarchy with the component hierarchy as we do this transformation, so if there are components that we didn't catch as we transform GDP stuff, we throw an error at the end since we may have inadvertently reclassified them. So your example above throws this error:
```
Traceback (most recent call last):
  File "/home/esjohn/src/pyomo/thing.py", line 10, in <module>
    TransformationFactory('gdp.transform_current_disjunctive_state').apply_to(m)
  File "/home/esjohn/src/pyomo/pyomo/core/base/transformation.py", line 77, in apply_to
    reverse_token = self._apply_to(model, **kwds)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/esjohn/src/pyomo/pyomo/gdp/plugins/transform_current_disjunctive_state.py", line 146, in _apply_to
    raise GDP_Error(
pyomo.gdp.disjunct.GDP_Error: Found active Disjuncts on the model that were not included in any Disjunctions:
d[3], d[2], e[2]
Please deactivate them or include them in a Disjunction.
```